File: backend/TirthSeva.API/Services/BhaktnivasService.cs
using Microsoft.EntityFrameworkCore;
using TirthSeva.API.Data;
using TirthSeva.API.DTOs;
using TirthSeva.API.Models;

namespace TirthSeva.API.Services
{
    public class BhaktnivasService
    {
        private readonly ApplicationDbContext _context;

        public BhaktnivasService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<List<BhaktnivasListDTO>> GetAllBhaktnivasAsync(int? templeId, decimal? minPrice, decimal? maxPrice, bool? isAvailable)
        {
            var query = _context.Bhaktnivas
                .Include(b => b.Temple)
                .Include(b => b.ServiceProvider)
                .AsQueryable();

            if (templeId.HasValue)
            {
                query = query.Where(b => b.TempleId == templeId.Value);
            }

            if (minPrice.HasValue)
            {
                query = query.Where(b => b.PricePerNight >= minPrice.Value);
            }

            if (maxPrice.HasValue)
            {
                query = query.Where(b => b.PricePerNight <= maxPrice.Value);
            }

            if (isAvailable.HasValue)
            {
                query = query.Where(b => b.IsAvailable == isAvailable.Value);
            }

            return await query
                .OrderBy(b => b.PricePerNight)
                .Select(b => new BhaktnivasListDTO
                {
                    Id = b.Id,
                    TempleId = b.TempleId,
                    TempleName = b.Temple.Name,
                    Name = b.Name,
                    PricePerNight = b.PricePerNight,
                    Capacity = b.Capacity,
                    IsAvailable = b.IsAvailable,
                    DistanceFromTemple = b.DistanceFromTemple,
                    ImageUrl = b.ImageUrl,
                    Latitude = b.Latitude,
                    Longitude = b.Longitude,
                    ServiceProviderName = b.ServiceProvider.Name
                })
                .ToListAsync();
        }

        public async Task<BhaktnivasDetailDTO?> GetBhaktnivasByIdAsync(int id)
        {
            var bhaktnivas = await _context.Bhaktnivas
                .Include(b => b.Temple)
                .Include(b => b.ServiceProvider)
                .FirstOrDefaultAsync(b => b.Id == id);

            if (bhaktnivas == null)
            {
                return null;
            }

            return new BhaktnivasDetailDTO
            {
                Id = bhaktnivas.Id,
                TempleId = bhaktnivas.TempleId,
                TempleName = bhaktnivas.Temple.Name,
                Name = bhaktnivas.Name,
                Description = bhaktnivas.Description,
                PricePerNight = bhaktnivas.PricePerNight,
                Capacity = bhaktnivas.Capacity,
                IsAvailable = bhaktnivas.IsAvailable,
                DistanceFromTemple = bhaktnivas.DistanceFromTemple,
                ImageUrl = bhaktnivas.ImageUrl,
                Latitude = bhaktnivas.Latitude,
                Longitude = bhaktnivas.Longitude,
                ServiceProviderId = bhaktnivas.ServiceProviderId,
                ServiceProviderName = bhaktnivas.ServiceProvider.Name,
                CreatedAt = bhaktnivas.CreatedAt
            };
        }

        public async Task<List<BhaktnivasListDTO>> GetBhaktnivasByProviderAsync(int providerId)
        {
            return await _context.Bhaktnivas
                .Include(b => b.Temple)
                .Include(b => b.ServiceProvider)
                .Where(b => b.ServiceProviderId == providerId)
                .OrderByDescending(b => b.CreatedAt)
                .Select(b => new BhaktnivasListDTO
                {
                    Id = b.Id,
                    TempleId = b.TempleId,
                    TempleName = b.Temple.Name,
                    Name = b.Name,
                    PricePerNight = b.PricePerNight,
                    Capacity = b.Capacity,
                    IsAvailable = b.IsAvailable,
                    DistanceFromTemple = b.DistanceFromTemple,
                    ImageUrl = b.ImageUrl,
                    Latitude = b.Latitude,
                    Longitude = b.Longitude,
                    ServiceProviderName = b.ServiceProvider.Name
                })
                .ToListAsync();
        }

        public async Task<Bhaktnivas?> CreateBhaktnivasAsync(CreateBhaktnivasRequest request, int serviceProviderId)
        {
            var bhaktnivas = new Bhaktnivas
            {
                TempleId = request.TempleId,
                ServiceProviderId = serviceProviderId,
                Name = request.Name,
                Description = request.Description,
                PricePerNight = request.PricePerNight,
                Capacity = request.Capacity,
                DistanceFromTemple = request.DistanceFromTemple,
                ImageUrl = request.ImageUrl,
                Latitude = request.Latitude,
                Longitude = request.Longitude,
                Amenities = request.Amenities,
                ContactPhone = request.ContactPhone,
                IsAvailable = true,
                CreatedAt = DateTime.UtcNow
            };

            _context.Bhaktnivas.Add(bhaktnivas);
            await _context.SaveChangesAsync();

            return bhaktnivas;
        }

        public async Task<Bhaktnivas?> UpdateBhaktnivasAsync(int id, UpdateBhaktnivasRequest request, int? serviceProviderId = null)
        {
            var bhaktnivas = await _context.Bhaktnivas.FindAsync(id);
            if (bhaktnivas == null)
            {
                return null;
            }

            // If service provider ID is provided, check ownership
            if (serviceProviderId.HasValue && bhaktnivas.ServiceProviderId != serviceProviderId.Value)
            {
                return null; // Not authorized
            }

            bhaktnivas.TempleId = request.TempleId;
            bhaktnivas.Name = request.Name;
            bhaktnivas.Description = request.Description;
            bhaktnivas.PricePerNight = request.PricePerNight;
            bhaktnivas.Capacity = request.Capacity;
            bhaktnivas.DistanceFromTemple = request.DistanceFromTemple;
            bhaktnivas.ImageUrl = request.ImageUrl;
            bhaktnivas.Latitude = request.Latitude;
            bhaktnivas.Longitude = request.Longitude;
            bhaktnivas.Amenities = request.Amenities;
            bhaktnivas.ContactPhone = request.ContactPhone;
            bhaktnivas.IsAvailable = request.IsAvailable;

            await _context.SaveChangesAsync();
            return bhaktnivas;
        }

        public async Task<bool> DeleteBhaktnivasAsync(int id, int? serviceProviderId = null)
        {
            var bhaktnivas = await _context.Bhaktnivas.FindAsync(id);
            if (bhaktnivas == null)
            {
                return false;
            }

            // If service provider ID is provided, check ownership
            if (serviceProviderId.HasValue && bhaktnivas.ServiceProviderId != serviceProviderId.Value)
            {
                return false; // Not authorized
            }

            // Delete associated bookings first to avoid foreign key constraint violations
            var bookings = await _context.Bookings.Where(b => b.BhaktnivasId == id).ToListAsync();
            _context.Bookings.RemoveRange(bookings);

            _context.Bhaktnivas.Remove(bhaktnivas);
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> UpdateAvailabilityAsync(int id, bool isAvailable, int serviceProviderId)
        {
            var bhaktnivas = await _context.Bhaktnivas.FindAsync(id);
            if (bhaktnivas == null || bhaktnivas.ServiceProviderId != serviceProviderId)
            {
                return false;
            }

            bhaktnivas.IsAvailable = isAvailable;
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<int?> GetAvailableCapacityAsync(int bhaktnivasId, DateTime checkIn, DateTime checkOut)
        {
            var slots = await _context.BhaktnivasSlots
                .Where(s => s.BhaktnivasId == bhaktnivasId && s.Date >= checkIn.Date && s.Date < checkOut.Date)
                .ToListAsync();

            var nightsCount = (checkOut.Date - checkIn.Date).Days;
            
            if (slots.Count < nightsCount) return 0; // Not all nights are released

            return slots.Min(s => s.AvailableCapacity);
        }

        public async Task<List<BhaktnivasSlotDTO>> GetSlotsAsync(int bhaktnivasId, DateTime? fromDate, DateTime? toDate)
        {
            var query = _context.BhaktnivasSlots
                .Where(s => s.BhaktnivasId == bhaktnivasId);

            if (fromDate.HasValue) query = query.Where(s => s.Date >= fromDate.Value.Date);
            if (toDate.HasValue) query = query.Where(s => s.Date <= toDate.Value.Date);

            return await query
                .OrderBy(s => s.Date)
                .Select(s => new BhaktnivasSlotDTO
                {
                    Id = s.Id,
                    BhaktnivasId = s.BhaktnivasId,
                    Date = s.Date,
                    TotalCapacity = s.TotalCapacity,
                    AvailableCapacity = s.AvailableCapacity
                })
                .ToListAsync();
        }

        public async Task<bool> ReleaseSlotsAsync(int bhaktnivasId, ReleaseBhaktnivasSlotsRequest request, int providerId)
        {
            var bhaktnivas = await _context.Bhaktnivas.FindAsync(bhaktnivasId);
            if (bhaktnivas == null || bhaktnivas.ServiceProviderId != providerId) return false;

            for (var date = request.StartDate.Date; date <= request.EndDate.Date; date = date.AddDays(1))
            {
                var existingSlot = await _context.BhaktnivasSlots
                    .FirstOrDefaultAsync(s => s.BhaktnivasId == bhaktnivasId && s.Date == date);

                if (existingSlot != null)
                {
                    // Update capacity if exist. 
                    // Warning: decrementing total capacity might lead to negative availability if already booked.
                    // For simplicity, we just update it.
                    existingSlot.TotalCapacity = request.Capacity;
                    // Reset available capacity to new total (this is a bit destructive but simple for now)
                    // Better logic: increment/decrement based on diff.
                    existingSlot.AvailableCapacity = request.Capacity; 
                }
                else
                {
                    _context.BhaktnivasSlots.Add(new BhaktnivasSlot
                    {
                        BhaktnivasId = bhaktnivasId,
                        Date = date,
                        TotalCapacity = request.Capacity,
                        AvailableCapacity = request.Capacity
                    });
                }
            }

            await _context.SaveChangesAsync();
            return true;
        }
    }
}

File: backend-spring/src/main/java/com/tirthseva/api/service/BhaktnivasService.java
package com.tirthseva.api.service;

import com.tirthseva.api.dto.bhaktnivas.*;
import com.tirthseva.api.entity.Bhaktnivas;
import com.tirthseva.api.entity.BhaktnivasSlot;
import com.tirthseva.api.repository.BhaktnivasRepository;
import com.tirthseva.api.repository.BhaktnivasSlotRepository;
import com.tirthseva.api.repository.BookingRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class BhaktnivasService {

    private final BhaktnivasRepository bhaktnivasRepository;
    private final BhaktnivasSlotRepository bhaktnivasSlotRepository;
    private final BookingRepository bookingRepository;

    public List<BhaktnivasListDTO> getAllBhaktnivas(Integer templeId, BigDecimal minPrice,
            BigDecimal maxPrice, Boolean isAvailable) {
        List<Bhaktnivas> bhaktnivasList = bhaktnivasRepository.findByFilters(templeId, minPrice, maxPrice, isAvailable);
        return bhaktnivasList.stream()
                .map(this::mapToListDTO)
                .collect(Collectors.toList());
    }

    public BhaktnivasDetailDTO getBhaktnivasById(Integer id) {
        Bhaktnivas bhaktnivas = bhaktnivasRepository.findById(id).orElse(null);
        return bhaktnivas != null ? mapToDetailDTO(bhaktnivas) : null;
    }

    public List<BhaktnivasListDTO> getBhaktnivasByProvider(Integer providerId) {
        return bhaktnivasRepository.findByServiceProviderIdOrderByCreatedAtDesc(providerId).stream()
                .map(this::mapToListDTO)
                .collect(Collectors.toList());
    }

    @Transactional
    public Bhaktnivas createBhaktnivas(CreateBhaktnivasRequest request, Integer serviceProviderId) {
        Bhaktnivas bhaktnivas = Bhaktnivas.builder()
                .templeId(request.getTempleId())
                .serviceProviderId(serviceProviderId)
                .name(request.getName())
                .description(request.getDescription() != null ? request.getDescription() : "")
                .pricePerNight(request.getPricePerNight())
                .capacity(request.getCapacity())
                .distanceFromTemple(request.getDistanceFromTemple() != null ? request.getDistanceFromTemple() : "")
                .imageUrl(request.getImageUrl() != null ? request.getImageUrl() : "")
                .latitude(request.getLatitude() != null ? request.getLatitude() : 0.0)
                .longitude(request.getLongitude() != null ? request.getLongitude() : 0.0)
                .amenities(request.getAmenities() != null ? request.getAmenities() : "")
                .contactPhone(request.getContactPhone() != null ? request.getContactPhone() : "")
                .isAvailable(true)
                .createdAt(LocalDateTime.now())
                .build();

        return bhaktnivasRepository.save(bhaktnivas);
    }

    @Transactional
    public Bhaktnivas updateBhaktnivas(Integer id, UpdateBhaktnivasRequest request, Integer serviceProviderId) {
        Bhaktnivas bhaktnivas = bhaktnivasRepository.findById(id).orElse(null);

        if (bhaktnivas == null) {
            return null;
        }

        // If serviceProviderId is provided, check ownership
        if (serviceProviderId != null && !bhaktnivas.getServiceProviderId().equals(serviceProviderId)) {
            return null;
        }

        bhaktnivas.setTempleId(request.getTempleId());
        bhaktnivas.setName(request.getName());
        bhaktnivas.setDescription(request.getDescription() != null ? request.getDescription() : "");
        bhaktnivas.setPricePerNight(request.getPricePerNight());
        bhaktnivas.setCapacity(request.getCapacity());
        bhaktnivas
                .setDistanceFromTemple(request.getDistanceFromTemple() != null ? request.getDistanceFromTemple() : "");
        bhaktnivas.setImageUrl(request.getImageUrl() != null ? request.getImageUrl() : "");
        bhaktnivas.setLatitude(request.getLatitude() != null ? request.getLatitude() : 0.0);
        bhaktnivas.setLongitude(request.getLongitude() != null ? request.getLongitude() : 0.0);
        bhaktnivas.setAmenities(request.getAmenities() != null ? request.getAmenities() : "");
        bhaktnivas.setContactPhone(request.getContactPhone() != null ? request.getContactPhone() : "");
        bhaktnivas.setIsAvailable(request.getIsAvailable() != null ? request.getIsAvailable() : true);

        return bhaktnivasRepository.save(bhaktnivas);
    }

    @Transactional
    public boolean deleteBhaktnivas(Integer id, Integer serviceProviderId) {
        Bhaktnivas bhaktnivas = bhaktnivasRepository.findById(id).orElse(null);

        if (bhaktnivas == null) {
            return false;
        }

        if (serviceProviderId != null && !bhaktnivas.getServiceProviderId().equals(serviceProviderId)) {
            return false;
        }

        // Delete associated bookings first
        bookingRepository.findByBhaktnivasId(id).forEach(bookingRepository::delete);

        bhaktnivasRepository.delete(bhaktnivas);
        return true;
    }

    @Transactional
    public boolean updateAvailability(Integer id, Boolean isAvailable, Integer serviceProviderId) {
        Bhaktnivas bhaktnivas = bhaktnivasRepository.findById(id).orElse(null);

        if (bhaktnivas == null || !bhaktnivas.getServiceProviderId().equals(serviceProviderId)) {
            return false;
        }

        bhaktnivas.setIsAvailable(isAvailable);
        bhaktnivasRepository.save(bhaktnivas);
        return true;
    }

    public Integer getAvailableCapacity(Integer bhaktnivasId, LocalDate checkIn, LocalDate checkOut) {
        List<BhaktnivasSlot> slots = bhaktnivasSlotRepository
                .findByBhaktnivasIdAndDateGreaterThanEqualAndDateLessThan(bhaktnivasId, checkIn, checkOut);

        long nightsCount = checkIn.until(checkOut).getDays();

        if (slots.size() < nightsCount) {
            return 0; // Not all nights are released
        }

        return slots.stream()
                .mapToInt(BhaktnivasSlot::getAvailableCapacity)
                .min()
                .orElse(0);
    }

    public List<BhaktnivasSlotDTO> getSlots(Integer bhaktnivasId, LocalDate fromDate, LocalDate toDate) {
        List<BhaktnivasSlot> slots;

        if (fromDate != null && toDate != null) {
            slots = bhaktnivasSlotRepository.findByBhaktnivasIdAndDateBetween(bhaktnivasId, fromDate, toDate);
        } else {
            slots = bhaktnivasSlotRepository.findByBhaktnivasIdOrderByDateAsc(bhaktnivasId);
        }

        return slots.stream()
                .map(s -> BhaktnivasSlotDTO.builder()
                        .id(s.getId())
                        .bhaktnivasId(s.getBhaktnivasId())
                        .date(s.getDate())
                        .totalCapacity(s.getTotalCapacity())
                        .availableCapacity(s.getAvailableCapacity())
                        .build())
                .collect(Collectors.toList());
    }

    @Transactional
    public boolean releaseSlots(Integer bhaktnivasId, ReleaseBhaktnivasSlotsRequest request, Integer providerId) {
        Bhaktnivas bhaktnivas = bhaktnivasRepository.findById(bhaktnivasId).orElse(null);

        if (bhaktnivas == null || !bhaktnivas.getServiceProviderId().equals(providerId)) {
            return false;
        }

        LocalDate currentDate = request.getStartDate();
        while (!currentDate.isAfter(request.getEndDate())) {
            BhaktnivasSlot existingSlot = bhaktnivasSlotRepository
                    .findByBhaktnivasIdAndDate(bhaktnivasId, currentDate)
                    .orElse(null);

            if (existingSlot != null) {
                existingSlot.setTotalCapacity(request.getCapacity());
                existingSlot.setAvailableCapacity(request.getCapacity());
                bhaktnivasSlotRepository.save(existingSlot);
            } else {
                BhaktnivasSlot newSlot = BhaktnivasSlot.builder()
                        .bhaktnivasId(bhaktnivasId)
                        .date(currentDate)
                        .totalCapacity(request.getCapacity())
                        .availableCapacity(request.getCapacity())
                        .build();
                bhaktnivasSlotRepository.save(newSlot);
            }

            currentDate = currentDate.plusDays(1);
        }

        return true;
    }

    private BhaktnivasListDTO mapToListDTO(Bhaktnivas b) {
        return BhaktnivasListDTO.builder()
                .id(b.getId())
                .templeId(b.getTempleId())
                .templeName(b.getTemple() != null ? b.getTemple().getName() : "")
                .name(b.getName())
                .pricePerNight(b.getPricePerNight())
                .capacity(b.getCapacity())
                .isAvailable(b.getIsAvailable())
                .distanceFromTemple(b.getDistanceFromTemple())
                .imageUrl(b.getImageUrl())
                .latitude(b.getLatitude())
                .longitude(b.getLongitude())
                .serviceProviderName(b.getServiceProvider() != null ? b.getServiceProvider().getName() : "")
                .build();
    }

    private BhaktnivasDetailDTO mapToDetailDTO(Bhaktnivas b) {
        return BhaktnivasDetailDTO.builder()
                .id(b.getId())
                .templeId(b.getTempleId())
                .templeName(b.getTemple() != null ? b.getTemple().getName() : "")
                .name(b.getName())
                .description(b.getDescription())
                .pricePerNight(b.getPricePerNight())
                .capacity(b.getCapacity())
                .isAvailable(b.getIsAvailable())
                .distanceFromTemple(b.getDistanceFromTemple())
                .imageUrl(b.getImageUrl())
                .latitude(b.getLatitude())
                .longitude(b.getLongitude())
                .serviceProviderId(b.getServiceProviderId())
                .serviceProviderName(b.getServiceProvider() != null ? b.getServiceProvider().getName() : "")
                .createdAt(b.getCreatedAt())
                .build();
    }
}
